### PR TITLE
docs: add/modify specs to describe auction types

### LINKED ISSUE #84: docs: add/modify specs to describe auction types
## Summary
This is to add and modify the spec documents to describe auction types of fixed price auction and batch auction.

## Tasks

- [ ] Draft 01_concept.md

- [ ] Draft 02_state.md

- [ ] Draft 03_state_transitions.md

- [ ] Draft 04_messages.md 

- [ ] Draft 05_end_block.md

- [ ] Draft 06_events.md

- [ ] Draft 07_params.md

- [ ] Clarify the explanations

- [ ] Modify the related proto files to define the states

## References
- Continued work from https://github.com/tendermint/fundraising/issues/81
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by jeonghoyeo7**
## Description

This is to add the details of specs for the module.

closes: #84

## Tasks

- [x] Draft 01_concept.md
- [x] Draft 02_state.md
- [x] Draft 03_state_transitions.md
- [x] Draft 04_messages.md 
- [x] Draft 05_end_block.md
- [x] Modifying related .proto files and related refactoring

## References
- Continued work from https://github.com/tendermint/fundraising/issues/81

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://go.dev/blog/godoc).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/fundraising/fundraising.proto [hunk 1]
```diff
@@ -94,22 +94,25 @@ message BaseAuction {
     (gogoproto.nullable) = false
   ];
 
+  // current number of winning bidders
+  uint64 number_winning_bidders = 13;
```
Comment (jaybxyz) [inline proto/fundraising/fundraising.proto]: How should we handle the value for `FixedPriceAuction`? Should we increment by one when a bid gets success?

File: proto/fundraising/fundraising.proto [hunk 2]
```diff
@@ -121,27 +124,21 @@ message FixedPriceAuction {
...
 
-// EnglishAuction defines the english auction type. It is an ascending dynamic
-// auction that an auctioneer decides the starting price of the selling amount
-// of coin and bidders bid to purchase the amounts of coin. It creates
-// competition for the price, not how fast bidders can bid.
-message EnglishAuction {
+// BatAuction defines a batch auction type. It allows each bidder
+// to participate in the auction by placing limit orders with a bid price
+// chosen freely at any time within the auction period. An order book is created
+// to record the bids with various bid prices.
```
Comment (jaybxyz) [inline proto/fundraising/fundraising.proto]: A little typo in `BatAuction` and I add more context around the `BatchAuction`. 
Feel free to add/update the suggested lines.
```suggestion
// BatchAuction defines a batch auction type. It allows bidders to participate in the auction 
// by placing their limit orders with a bid price they are willing to bid within the auction period.
// They can place multiple bids with different bid prices and if they want to modify their existing bid, 
// they only have an option to modify with a higher bid price. Under the hood, an order book is created 
// to record the bids to calculate winners.
```

File: proto/fundraising/fundraising.proto [hunk 3]
```diff
@@ -121,27 +124,21 @@ message FixedPriceAuction {
...
-  uint32 extended = 3;
+  // max_extended_round specifies a maximum number of extended rounds
+  uint32 max_extended_round = 2;
 
-  // extend_rate specifies the rate that decides if the auction needs another
-  // round
-  string extend_rate = 4 [
+  // rate_extended_round specifies the rate that decides if the auction needs
+  // another round
+  string rate_extended_round = 3 [
```
Comment (jaybxyz) [inline proto/fundraising/fundraising.proto]: I think the name should be `extended_round_rate`.

File: x/fundraising/spec/02_state.md [hunk 4]
```diff
@@ -46,7 +46,12 @@ type AuctionI interface {
 
 	GetVestingSchedules() []VestingSchedule
 	SetVestingSchedules([]VestingSchedule) error
-
+	
+	GetWinningPrice() sdk.Dec  // To load WinningPrice at the current time
+	GetNumberWinningBidders() uint64
+	
+	GetRemainingSellingCoin() sdk.Coin // To calculate RemainingSellingCoin
```
Comment (jaybxyz) [inline x/fundraising/spec/02_state.md]: Is there a reason why you don't have Setter functions?

**Review by jaybxyz — CHANGES_REQUESTED** [review]
I think we need to update `tx.proto` as well. Also, I see a number of places where `EnglishAuction` is used in comments and docs.

File: proto/fundraising/fundraising.proto (hunk 1 above)
Comment (jeonghoyeo7) [reply proto/fundraising/fundraising.proto]: Yes it can be one way.

File: proto/fundraising/fundraising.proto (hunk 2 above)
Comment (jeonghoyeo7) [reply proto/fundraising/fundraising.proto]: I will reflect this.

File: proto/fundraising/fundraising.proto (hunk 3 above)
Comment (jeonghoyeo7) [reply proto/fundraising/fundraising.proto]: I will reflect this.

File: x/fundraising/spec/02_state.md (hunk 4 above)
Comment (jeonghoyeo7) [reply x/fundraising/spec/02_state.md]: I will add the setters.

**Comment by jeonghoyeo7** [comment]
- I've changed based on the comments.
 - Also changed the remaining "English" related.
 - "sequence" --> "BidId", "Eligible" --> "isWInner"

But, it seems it needs to add AllowedBidders to MsgCreate...Auction in msg.go

**Comment by jaybxyz** [comment]
>But, it seems it needs to add AllowedBidders to MsgCreate...Auction in msg.go


All auctions are closed when they are created. The fundraising module is designed to delegate the control to an external module.

File: proto/fundraising/fundraising.proto [hunk 5]
```diff
@@ -239,11 +242,11 @@ message Bid {
   string bidder = 2;
 
   // id specifies an index of a bid for the bidder
-  uint64 id = 3;
+  uint64 bid_id = 3;
```
Comment (jaybxyz) [inline proto/fundraising/fundraising.proto]: Since id is already inside `Bid` struct, I think it can just be `id`.

**Review by jaybxyz — CHANGES_REQUESTED** [review]
I have a couple things that are not covered in this PR.

1. `NewBaseAuction` needs to have an added argument(s). Update the affected codebase.
     - `NumberWinningBidders` 
2.  We can follow SDK convention of using shorten word for `Number`, so I would change from `NumberWinningBidders` to `NumWinningBidders` (It is shorten for a number of winning bidders)
3. I found a number of places that are using `sequence`. They need to be updated.

File: proto/fundraising/fundraising.proto (hunk 5 above)
Comment (jeonghoyeo7) [reply proto/fundraising/fundraising.proto]: I updated based on the comments 1,2,3 above.
For adding NumWinningBidders to CreateFixedPriceAuction of keeper/auction.go, I added NumWinningBidders = 0.

**Comment by jaybxyz** [comment]
Once you merge from the latest `main` branch and resolve conflicts, I will approve this PR.

**Review by jaybxyz — APPROVED** [review]

**Pull request merged**